### ISSUE TITLE
Add a secrets handling mechanism

### ISSUE THREAD

**Issue by johananl** [kinvolk/lokomotive#580]
## General

Lokomotive doesn't currently have an official way for handling secrets. Secrets are specified as either environment variables or in the config file, and it's up to the user to figure out how to store the secrets, retrieve them securely etc.

We should consider adding a "smarter" way for handling secrets. For a start, I suggest we look into integrating with the following secrets engines:

- [Vault](https://www.vaultproject.io/)
- [AWS SSM](https://docs.aws.amazon.com/systems-manager/latest/userguide/systems-manager-parameter-store.html)

Related issue: https://github.com/kinvolk/lokomotive/issues/466

## Implementation

The secrets handling mechanism should ideally be implemented as a **generic layer** which abstracts the underlying providers. Following is a **very** general idea of how to implement this.

```hcl
secrets_provider "vault" {
  url = "https://vault.example.com:8200"
  client_cert_file = "ca.crt"
}

secret "stuff_password" {
  provider = secrets_provider.vault
  path = "secret/prod/stuff"
}

cluster "packet" {
  ...
  component "stuff" {
    password = secret.stuff_password.value
  }
}
```

Things to note:

- `secrets_provider` is a new HCL resource to be defined. In the example above a provider of type `vault` is used. Other providers may have different arguments. This is fine and expected.
- `secret` is a new HCL resource to be defined. This should ideally be a **generic** resource, so we should check if it's possible to specify the "path" for getting a secret in a way which could work for multiple secrets providers.
- `secret` exposes an **attribute** (similar to Terraform attributes) named `value` which returns the clear-text value of a secret after it was retrieved from the secrets provider.
- Users **define** secrets in the top-level of the cluster config, then **reference** them within `cluster` or `component` sections as `secret.<name>.value` (maybe we don't even need `value` - need to check).

**Comment by johananl** [comment]
@iaguis @schu - FYI.

**Comment by johananl** [comment]
Decided to run a timeboxed PoC to figure out if it's useful to have this feature and if it can be implemented in a reasonable manner.

Concerns from the team:

- Maintenance burden of supporting various secrets backends.
- "Unix philosophy" - maybe we should keep secrets out of Lokomotive and let the users solve the problem on their own and interpolate the values into the cluster config however they see fit.

**Comment by invidian** [comment]
From my perspective, proposed implementation looks very similar to the variables system we have right now. We take a value from **somewhere** and we inject it into out blocks configuration. So I'm wondering, if handling secrets should be much different from handling regular variables.

As far as I understand, we don't plan any **security** oriented features around those secrets (like I don't know, removing them from the memory right after using or something), so maybe rather than adding new HCL type for handling secrets, we could somehow extend the variables handling to be able to pull the values from some other places, like mentioned Vault? For example, make a default provider for variables a `lokocfg` file, but also support other providers, something like:
```hcl
variable_provider "vault" {
  url = "https://vault.example.com:8200"
  client_cert_file = "ca.crt"
}

variable "stuff_password" {
  provider = "vault"
  path = "secret/prod/stuff"
}
```

I support a concern regarding the maintenance of various backends, as there will always be a user asking  "Why backend X is not supported?". With such features, we must be careful to avoid situation, when we need to maintain a lot of code or we end up being bad maintainers, favoring some solution over others. Of course, if users of this feature grows, we can decide to support only most popular backends and I think this is then justified, that we won't support some of them.

Also, to mitigate this issue, 2 solutions comes to my mind:
- Support Go-based plugin system, similar to what bootkube is doing right now (https://github.com/kubernetes-sigs/bootkube/pull/1081). We could then provide and maintain set of built-in providers, while others could be dynamically loaded via the plugin file.
- Support `shell` provider, which would allow executing arbitrary commands to get values (perhaps from `stdin`). This way, we give users a freedom of solution to choose.

#

#### Alternative solution

Instead of adding secrets handling mechanism, we could document our recommended way of handling secrets, let's say with `blackbox` and eventually polish some rough edges with our implementation, which would better support it.

For example, while working with Terraform, one can use `blackbox` or `git-secret` to encrypt `secrets.auto.tfvars` file, which will only contain secrets, while keeping `variables.auto.tfvars` unencrypted for non-sensitive values. Then, before running Terraform and after modifying secrets, some shell commands needs to be executed to unveil/re-encrypt the secrets.

**Comment by johananl** [comment]
@invidian yes, secrets are similar to variables. This was my intent exactly when proposing the new `secret` type.
I don't see the advantage of "extending" the HCL `variable` directive. Why is it better then adding a new type? One reason *not* to tamper with `variable` is that if we do that we practically become maintainers of this part of the HCL library (right?).

>I support a concern regarding the maintenance of various backends, as there will always be a user asking "Why backend X is not supported?".

AFAICT this is true for nearly any feature: supported platforms, supported DNS providers, supported ingress controllers etc. It's our job as maintainers to choose what to maintain, and indeed it usually makes sense to start with one or two very popular options. Sure, having an "extensible" mechanism could make sense, but I'd say this may be thinking too far. I want to start simple and see if it makes sense to even have a secrets mechanism in Lokomotive.

>Instead of adding secrets handling mechanism, we could document our recommended way of handling secrets, let's say with blackbox and eventually polish some rough edges with our implementation, which would better support it.

If we decide not to solve the secrets handling problem in Lokomotive, I agree we should document our recommendation for handling secrets around the cluster config. However, it doesn't make sense to me to officially recommend Blackbox or any other similar solution because these tools have nothing to do with Lokomotive. To emphasize my point, we could also recommend to the user which shell or terminal emulator to use when managing Lokomotive clusters, however that would likely be weird :confused: 

To summarize my view, I'd say we either implement proper secrets handling in `lokoctl`, or we document **guidelines** on how to handle secrets (but not advocate for a specific tool).

**Comment by invidian** [comment]
>@invidian yes, secrets are similar to variables. This was my intent exactly when proposing the new secret type.
I don't see the advantage of "extending" the HCL variable directive. Why is it better then adding a new type? One reason not to tamper with variable is that if we do that we practically become maintainers of this part of the HCL library (right?).

We implement `variable` type already, here: https://github.com/kinvolk/lokomotive/blob/master/pkg/config/config.go#L32. So it would not be extending or maintaining HCL library.

I think the main advantage of **transparently extending existing type** is preserving the same "output" user interface, so `var.grafana_password` stays the same, but it has a new, different source.

>AFAICT this is true for nearly any feature: supported platforms, supported DNS providers, supported ingress controllers etc. It's our job as maintainers to choose what to maintain, and indeed it usually makes sense to start with one or two very popular options. Sure, having an "extensible" mechanism could make sense, but I'd say this may be thinking too far. I want to start simple and see if it makes sense to even have a secrets mechanism in Lokomotive.

You're right. This seems to be thinking too far right now. But we should definitely make sure that the implementation is easy to extend in the future in such case.

>If we decide not to solve the secrets handling problem in Lokomotive, I agree we should document our recommendation for handling secrets around the cluster config. However, it doesn't make sense to me to officially recommend Blackbox or any other similar solution because these tools have nothing to do with Lokomotive. To emphasize my point, we could also recommend to the user which shell or terminal emulator to use when managing Lokomotive clusters, however that would likely be weird confused

> To summarize my view, I'd say we either implement proper secrets handling in lokoctl, or we document guidelines on how to handle secrets (but not advocate for a specific tool).

We could still link to/recommend some tools which in our opinion as a maintainers play well with Lokomotive configuration, for example ones which we use ourselves.
